File: backend/app/models/storage.py
from typing import List, Literal, Optional
from pydantic import BaseModel, Field, model_validator


class StorageItem(BaseModel):
    name: str = Field(..., description="Folder/file name")
    path: str = Field(..., description="Relative path from base")
    type: str = Field(..., description="'directory' or 'file'")
    size_bytes: int = Field(..., description="Size in bytes")
    size_mb: float = Field(..., description="Size in MB")
    file_count: Optional[int] = Field(None, description="Number of files (for directories)")
    children: Optional[List['StorageItem']] = Field(None, description="Subdirectories/files")
    modified_time: Optional[str] = Field(None, description="Last modified time")


class StorageStats(BaseModel):
    total_size_bytes: int
    total_size_mb: float
    total_size_gb: float
    total_files: int
    total_directories: int


class DiskUsage(BaseModel):
    total_gb: float
    used_gb: float
    free_gb: float
    used_percent: float
    mount_point: str


class CleanupConfig(BaseModel):
    enabled: bool = False
    dry_run: bool = True
    schedule_mode: Literal["daily", "hourly"] = "daily"
    schedule_hour: int = Field(2, ge=0, le=23)
    schedule_minute: int = Field(0, ge=0, le=59)
    interval_hours: int = Field(6, ge=1, le=24)
    trigger_usage_percent: float = Field(70.0, ge=10.0, le=99.0)
    target_usage_percent: float = Field(60.0, ge=5.0, le=98.0)
    min_images_per_folder: int = Field(25, ge=20, le=30)
    keep_recent_dates_per_recipe: int = Field(3, ge=1, le=30)

    @model_validator(mode="after")
    def _check_thresholds(self) -> "CleanupConfig":
        if self.target_usage_percent >= self.trigger_usage_percent:
            raise ValueError(
                "target_usage_percent must be lower than trigger_usage_percent"
            )
        return self


class CleanupRunResult(BaseModel):
    ran_at: str
    trigger: Literal["scheduled", "manual", "threshold-skip"]
    dry_run: bool
    disk_used_percent_before: float
    disk_used_percent_after: float
    files_deleted: int = 0
    folders_deleted: int = 0
    freed_mb: float = 0.0
    leaf_folders_scanned: int = 0
    duration_seconds: float = 0.0
    message: str = ""


StorageItem.model_rebuild()

File: frontend-ts/src/types/storage.ts
export interface StorageItem {
  name: string;
  path: string;
  type: 'directory' | 'file';
  size_bytes: number;
  size_mb: number;
  file_count?: number;
  children?: StorageItem[] | null;
  modified_time?: string;
}

export interface StorageStats {
  total_size_bytes: number;
  total_size_mb: number;
  total_size_gb: number;
  total_files: number;
  total_directories: number;
}

export interface DiskUsage {
  total_gb: number;
  used_gb: number;
  free_gb: number;
  used_percent: number;
  mount_point: string;
}

export type CleanupScheduleMode = 'daily' | 'hourly';

export interface CleanupConfig {
  enabled: boolean;
  dry_run: boolean;
  schedule_mode: CleanupScheduleMode;
  schedule_hour: number;
  schedule_minute: number;
  interval_hours: number;
  trigger_usage_percent: number;
  target_usage_percent: number;
  min_images_per_folder: number;
  keep_recent_dates_per_recipe: number;
}

export type CleanupTrigger = 'scheduled' | 'manual' | 'threshold-skip';

export interface CleanupRunResult {
  ran_at: string;
  trigger: CleanupTrigger;
  dry_run: boolean;
  disk_used_percent_before: number;
  disk_used_percent_after: number;
  files_deleted: number;
  folders_deleted: number;
  freed_mb: number;
  leaf_folders_scanned: number;
  duration_seconds: number;
  message: string;
}
